### PR TITLE
add new vpp features

### PR DESCRIPTION
**Pull request by xfengcarl**
new features:Deblocking filter, FRC filter, Total Color Correction, Non-Linear Anamorphic Scaling.Image Stabilization...
new color standard: sRGB stRGB, xvYCC601, xvYCC709, BT.2020

Signed-off-by: xfengcarl <carl.zhang@intel.com>

### CODE REVIEW

**Comment by xfengcarl** [comment]
change base to v2.0

File: va/va_vpp.h [hunk 1]
```diff
@@ -354,6 +462,73 @@ typedef struct _VAProcPipelineCaps {
...
+     * the corresponding bit is set in \ref rotation_flags. See
+     * "Rotation angles" for a description of rotation angles.
+     *
+     * A value of 0 means the underlying hardware does not support any
+     * rotation. Otherwise, a check for a specific rotation angle can be
+     * performed as follows:
+     *
+     * \code
+     * VAProcPipelineCaps pipeline_caps;
+     * ...
```
Comment (mypopydev) [inline va/va_vpp.h]: Dos chars?

File: va/va_vpp.h [hunk 2]
```diff
@@ -643,6 +989,56 @@ typedef struct _VAProcFilterCapColorBalance {
...
+/** \brief Capabilities specification for the Total Color Correction filter. */
+typedef struct _VAProcFilterCapTotalColorCorrection {
+    /** \brief Color to correct. */
+    VAProcTotalColorCorrectionType    type;
+    /** \brief Range of supported values for the specified color. */
+    VAProcFilterValueRange            range;
+} VAProcFilterCapTotalColorCorrection;
+
+/** \brief Capabilities specification for the Image Stabilization filter. */
+typedef struct _VAProcFilterCapImageStabilization {
```
Comment (mypopydev) [inline va/va_vpp.h]: Does it support in the UMD driver?

File: va/va_vpp.h [hunk 3]
```diff
@@ -643,6 +989,56 @@ typedef struct _VAProcFilterCapColorBalance {
...
+/** \brief Capabilities specification for the Frame Rate Conversion filter. */
+typedef struct _VAProcFilterCapFrameRateConversion { 
+    /** \brief Should be set to 1 if only supported rates are requested.
+        \brief Set to 0 to get the rest of the caps for the particular custom rate */
+    unsigned int                        bget_custom_rates;
+    /** \brief FRC custom rates supported by the pipeline in the first query
+        \brief App request caps for a custom rate in the second query */
+    unsigned int                        frc_custom_rates;
+    /** \brief FPS of input sequence. */
+    unsigned int                        input_fps;
```
Comment (mypopydev) [inline va/va_vpp.h]: How to handle the case when the input fps is 23.97 ?

File: va/va_vpp.h [hunk 4]
```diff
@@ -643,6 +989,56 @@ typedef struct _VAProcFilterCapColorBalance {
...
+    /** \brief FPS of output sequence. */
+    unsigned int                        output_fps;
+    /** \brief Number of input frames. */   
+    unsigned int                        input_frames;
+    /** \brief Number of output frames. */
+    unsigned int                        output_frames;
+   /** \brief Set to 1 if interlaced input is supoorted. */
+    unsigned int                        input_interlaced;
+} VAProcFilterCapFrameRateConversion;
+
```
Comment (xhaihao) [inline va/va_vpp.h]: Could you split the commit per feature?

File: va/va_vpp.h [hunk 5]
```diff
@@ -336,6 +428,22 @@ typedef enum _VAProcColorStandardType {
...
+/** \brief Chroma samples are co-sited vertically on the top with the luma samples. */
+#define VA_CHROMA_SITING_VERTICAL_TOP         0x00000001
+/** \brief Chroma samples are not co-sited vertically with the luma samples. */
+#define VA_CHROMA_SITING_VERTICAL_CENTER      0x00000002
+/** \brief Chroma samples are co-sited vertically on the bottom with the luma samples. */
+#define VA_CHROMA_SITING_VERTICAL_BOTTOM      0x00000003
+/** \brief Chroma samples are co-sited horizontally on the left with the luma samples. */
+#define VA_CHROMA_SITING_HORIZONTAL_LEFT      0x00000004
+/** \brief Chroma samples are not co-sited horizontally with the luma samples. */
+#define VA_CHROMA_SITING_HORIZONTAL_CENTER    0x00000008
```
Comment (xfengcarl) [inline va/va_vpp.h]: 1 horizontal and 1 vertical

File: va/va_vpp.h (hunk 2 above)
Comment (xfengcarl) [inline va/va_vpp.h]: yes

**Comment by xhaihao** [comment]
Could you please rebase your PR and change your PR to 'merge 24 commits to 01org:master'? All patches on v2.0-branch have been merged into master.

File: va/va.h [hunk 6]
```diff
@@ -238,6 +238,17 @@ typedef int VAStatus;	/** Return status type from functions */
 #define VA_PADDING_LARGE        32
 
 /**
+ * This is to indicate that the color-space conversion uses full range or reduced range.
+ * VA_SOURCE_RANGE_FULL(Full range): Y/Cb/Cr is in [0, 255]. It is mainly used
+ *      for JPEG/JFIF formats. The combination with the BT601 flag means that
+ *      JPEG/JFIF color-space conversion matrix is used.
+ * VA_SOURCE_RANGE_REDUCED(Reduced range): Y is in [16, 235] and Cb/Cr is in [16, 240].
+ *      It is mainly used for the YUV->RGB color-space conversion in SDTV/HDTV/UHDTV.
```
Comment (fhvwy) [inline va/va.h]: Can this be used to get limited range RGB as well?  (As used in HDMI.)

Maybe mention the ranges for 10-bit YUV in the comment, too.

File: va/va_vpp.h [hunk 7]
```diff
@@ -303,10 +306,77 @@ typedef enum _VAProcColorStandardType {
...
+    VAProcTotalColorCorrectionCount
+} VAProcTotalColorCorrectionType;
+/** @name Video blending flags */
+/**@{*/
+/** \brief Global alpha blending. */
+#define VA_BLEND_GLOBAL_ALPHA           0x0002
+/** \brief Premultiplied alpha blending (RGBA surfaces only). */
+#define VA_BLEND_PREMULTIPLIED_ALPHA    0x0008
+/** \brief Luma color key (YUV surfaces only). */
+#define VA_BLEND_LUMA_KEY               0x0010
```
Comment (fhvwy) [inline va/va_vpp.h]: Why do these values have gaps?  Are more going to be added later?

File: va/va_vpp.h [hunk 8]
```diff
@@ -303,10 +306,77 @@ typedef enum _VAProcColorStandardType {
...
+    /** \brief sRGB. */
+    VAProcColorStandardSRGB,
+    /** \brief stRGB. */
+    VAProcColorStandardSTRGB,
+    /** \brief xvYCC601. */
+    VAProcColorStandardXVYCC601,
+    /** \brief xvYCC709. */
+    VAProcColorStandardXVYCC709,
+    /** \brief ITU-R BT.2020. */
+    VAProcColorStandardBT2020,
```
Comment (fhvwy) [inline va/va_vpp.h]: Can you specify exactly what each of these mean?  (See the changes I added to the other values of this enum in #127 .)

In particular, I'm wondering what the BT2020 case actually is here - e.g. is it BT.2020 colour space, with the usual SDR transfer characteristic, and using BT.2020 non-constant luminance matrix coefficeints?

File: va/va_vpp.h [hunk 9]
```diff
@@ -336,6 +406,21 @@ typedef enum _VAProcColorStandardType {
 #define VA_PIPELINE_FLAG_END		0x00000004
 /**@}*/
 
+/** @name Chroma Siting flag */
+/**@{*/
+#define VA_CHROMA_SITING_UNKNOWN              0x00000000
```
Comment (fhvwy) [inline va/va_vpp.h]: This presumably actually means `VERTICAL_CENTRE | HORIZONTAL_LEFT` (i.e. the default for most video codecs) in all real cases?  Maybe just say that.

File: va/va_vpp.h (hunk 5 above)
[the remark this answers is not in the document]
Comment (fhvwy) [reply va/va_vpp.h]: Some extra values to give the masks and shifts might be nice.  Using them in defining these values would also make it clear that `HORIZONTAL_LEFT` and `HORIZONTAL_CENTRE` can't be combined despite looking like flags.

File: va/va_vpp.h [hunk 10]
```diff
@@ -355,8 +440,78 @@ typedef struct _VAProcPipelineCaps {
...
+    uint32_t        max_output_width;
+    /** \brief Max supported output height in pixels. */
+    uint32_t        max_output_height;
+    /** \brief Min supported output width in pixels. */
+    uint32_t        min_output_width;
+    /** \brief Min supported output height in pixels. */
+    uint32_t        min_output_height;
     /** \brief Reserved bytes for future use, must be zero */
-    uint32_t                va_reserved[VA_PADDING_LARGE];
+    #if defined(__AMD64__) || defined(__x86_64__) || defined(__amd64__)
```
Comment (fhvwy) [inline va/va_vpp.h]: This is going to go wrong on 64-bit non-x86 platforms.  Does `__LP64__` work everywhere we need to support?

Comment (xfengcarl) [reply va/va_vpp.h]: yes, will add it

File: va/va_vpp.h (hunk 5 above)
[the remark this answers is not in the document]
Comment (xfengcarl) [reply va/va_vpp.h]: horizontal should always combined with vertical, will change the comments to clear it.

File: va/va.h (hunk 6 above)
Comment (xfengcarl) [reply va/va.h]: I suppose we can add more color range, but I wonder whether we should add them now? who are using it?

File: va/va_vpp.h (hunk 8 above)
Comment (xfengcarl) [reply va/va_vpp.h]: yes, will refer #127

File: va/va.h (hunk 6 above)
Comment (jackie-guo) [reply va/va.h]: I think Full range and Reduced range can be used for both RGB and YUV and also 8-bit/10-bit.

File: va/va_vpp.h (hunk 9 above)
Comment (jackie-guo) [reply va/va_vpp.h]: In driver, if application does not specify chromasiting information, driver will use left/center as default chromasiting.

File: va/va_vpp.h (hunk 7 above)
Comment (xfengcarl) [reply va/va_vpp.h]: VA_BLEND_LUMA_KEY can OR'd with VA_BLEND_XXX_ALPHA

File: va/va_vpp.h [hunk 11]
```diff
@@ -374,6 +543,14 @@ typedef struct _VAProcFilterValueRange {
...
 } VAProcFilterValueRange;
 
+typedef struct _VAProcColorProperties { 
+    /** Chroma sample location.\c VA_CHROMA_SITING_VERTICAL_XXX | VA_CHROMA_SITING_HORIZONTAL_XXX */ 
+    uint8_t chroma_sample_location; 
+    /** Chroma sample location. \c VA_SOURCE_RANGE_XXX*/ 
+    uint8_t color_range; 
+    uint8_t reserved[6]; 
+} VAProcColorProperties; 
+
```
Comment (xfengcarl) [inline va/va_vpp.h]: refer  https://github.com/01org/libva/pull/127

**Pull request merged**